### ISSUE TITLE
hostname -s considered harmful?

### ISSUE THREAD

**Issue by ndokos** [distributed-system-analysis/pbench#436]
@vlaad ran into a problem when postprocessing where files were reported to be existing already. He and @mffiedler investigated and pinpointed the trouble to the fact that the remotes had their IP addresses as hostnames. pbench uses `hostname -s' to set $hostname which is then used by pbench-postprocess-tools: on the remotes, the results are moved down to $hostname (or $label:$hostname if label is defined); on the controller, the same thing happens (if tools are registered locally), and then the remote tool directories are copied over and deleted from the remotes. Unfortunately, all the remotes have names like 10.9.0.X, and hostname -s returns "10" for all of them.

One obvious solution is "don't do that": make sure that hostnames are unique (e.g. IP X.Y.Z.W generates a name like ip-X-Y-Z-W.domain, so hostname -s would return a unique name). It seems that that is not always possible: large-scale deployments have to accept whatever the deployed software dictates.

Another is to replace `hostname -s' by `hostname -f' in /opt/pbench-agent/base. This would resolve this situation, at the cost of using fqdns in the "normal" case, making directory paths even longer than they are. But this seems like the least painful solution. We could make it user-selectable if people complain.

Thoughts?

**Comment by atheurer** [comment]
I feel strongly that hostnames should be unique.  In what deployments is this not true?
Using fqdns is probably OK for me but what guarantees that is unique, if the hostname -s is not unique?

**Comment by mffiedler** [comment]
I've experienced this on OpenStack clusters where /etc/hostname contains IP
addresses instead of the nova instance name - at least until first reboot
of the instance after initial bringup.   hostname -s returns 10 or 172 or
192 on all the systems in the env.   Unique hostnames would be ideal, but
are there cases where this might be out of tester control?

Your question on fqdns is fair - no guarantee of uniqueness there either.

Regards,
Mike Fiedler
Red Hat OpenShift

**Comment by ndokos** [comment]
The fqdn solution would solve this particular problem (IP addresses used as hostnames), but indeed it would not guarantee uniqueness.

**Comment by atheurer** [comment]
Can we just enforce the hostname is whatever the user specified when calling register-tool --remote=<this_hostname>?  So if you use IP, short hostname, or fqdn at the time of registration, that is always used regardless of what hostname [-s] return on that system?

**Comment by ndokos** [comment]
Yes, we can - pbench-postprocess-tools would have to be modified to leave the remote tool data alone (instead of moving them to a $(hostname -s) subdir); the controller will do the pulling and moving into subdirs using the registered remote hostnames. IOW, the local case will have to be special-cased: on the remotes, nothing is done; but on the controller (assuming it has registered tools), we need to move its tool data down to a subdir. There's still the problem of the controller name: we'll have to use something like hostname -s or hostname -f here and hope for the best.

**Comment by mffiedler** [comment]
I think this proposed approach closes the window opening most of the way
with the added benefit of ensuring that the results subdirs use the
registered remote names.  I occasionally see problems where split data
shows up on the results server - some under the ip address and some under
the hostname.    Seems to happen when I register the remotes with ip addrs,
but some data gets put under a $(hostname -s) dir on the final results
server.  Registering by ip addrs is a convenience in large scale envs with
sequential ips.

> There's still the problem of the controller name: we'll have to use
something like hostname -s or hostname -f here and hope for the best.

Getting your controller hostname correct is a very reasonable
expectation/requirement.

Regards,
Mike Fiedler
Red Hat OpenShift
